### ISSUE TITLE
The `<model>` element is not compatible with modern web dev paradigms.

### ISSUE THREAD

**Issue by trusktr** [immersive-web/model-element#73]
Hello! I first wrote about this here.

https://twitter.com/trusktr/status/1717795902272974942

I want to open a conversation on how we could make the `<model>` element more compatible with today's web frameworks, and how currently it is not.

> [!Note]
> I work on [Lume, HTML elements for 3D rendering](https://lume.io), and I have been spent a lot of time imagining what native 3D elements in browsers could be like.
> Try viewing and hovering on elements in the devtools element inspector when view examples.

I'll re-post what I wrote on Twitter here, updated and improved a little:

---

@apple's HTML `<model>` element in [#visionOS](https://twitter.com/hashtag/visionOS?src=hashtag_click) Safari is incredibly neat (see [the video](https://developer.apple.com/videos/play/wwdc2023/10279/), at 12:36), hinting at an awesome future of high-level 3D web capabilities that don't require intimate programming knowledge to use.

It however does not hint at a full HTML 3D future compatible with React, Vue, Svelte, Solid, and all the other DOM tools we've evolved over the years. Let me explaine why, and perhaps we can fix this.

`<model>` enables amazing native features for web in a simple way, and starts to paint an awesome future for web, but I believe we can plan our HTML future better based on the present (cc @jensimmons as the curator of our [#webdev](https://twitter.com/hashtag/webdev?src=hashtag_click) needs for @WebKit, and @marcoscaceres as editor of the spec):

The current `model.getCamera()` API returns a non-HTMLElement plain-JavaScript object, which means all of our DOM-manipulation tools ([@buildWithLit](https://twitter.com/buildWithLit), [@reactjs](https://twitter.com/reactjs), [@vuejs](https://twitter.com/vuejs), [@sveltejs](https://twitter.com/sveltejs), [@solid_js](https://twitter.com/solid_js), and more) will not be able to manipulate a `<model>` element's camera out of the box using their declarative-reactive templating systems, requiring a non-standard new way of programming interactivity specifically for `<model>`. Web devs will build unnecessary wrappers (for every framework!) to bridge this gap, in order to tie `<model>` camera interactivity back into our paradigms.

As an example of the unnecessary wrappers people would need to create for all frameworks, [`react-three-fiber`](https://docs.pmnd.rs/react-three-fiber) wraps Three.js to add the declarative-reactive support for Three.js in React because Three.js is not DOM-based. The wrappers that people make for `<model>` will need to map attributes/properties from their templates to the `<model>`'s non-DOM camera objects.

With HTML elements (and well-designed Custom Elements), there is no need for wrappers to be made for every framework, all elements typically work out of the box in any web framework (there are a few edge cases, but we've learned to avoid those in design of new HTML elements).

The following HTML interface is an example that would be compatible with the DOM future all frameworks have been building:

```html
<model ...>
  <camera type="perspective" position="10 20 30"></camera>
</model>
```

This is compatible because by making the camera configurable via DOM with HTML support, every web framework can manipulate it out of the box (_and not just on the client side, but also on the server side!_).

Alternatively, camera attributes on the `<model>` element would also be a compatible pattern, and could later be superceded with HTML elements that override the attribute behavior:

```html
<model camera="perspective" camera-position="10 20 30" etc></model>
```

These designs are compatible with all of today's frameworks because they embrace the DOM tree model that all DOM tools are designed to work with.

The `getCamera()` API, though extremely nifty for what it can do, is not compatible because it side-steps the DOM trees that these tools manipulate, which will make our code [more WET (less DRY)](https://www.softwareyoga.com/is-your-code-dry-or-wet/#:~:text=Don%27t%20Repeat%20Yourself(DRY,t%20adhere%20to%20DRY%20principle.) due to those wrappers we'll need for every framework to connect to `<model>` element features (cameras). If the API were to give us elements and attributes (which elements, and which attributes, can be debated), we'd get these connections concisely and for free.

Imagine `<model>` comes out in all browsers, and out-of-the-box anyone using any framework can manipulate all aspects of a `<model>` scene right away without escaping out of their framework to write custom JavaScript.
 
With a DOM-based API, the `Promise` return would no longer be necessary for `getCamera()`, as these plain JS objects will be controlled while their rendering effects are abstracted underneath. The camera DOM can internally proxy to an underlying renderer (out of process or not).

Today, most frameworks practice one-way data flow (except in edge cases as with `<input>`s, etc), but the `getCamera()` API, besides not being compatible with DOM tools, is a more cumbersome two-way data flow design:

1. First, the user has to get a `Promise` and wait for it. With DOM API, users never have to wait for sub-objects (children) to be ready by using Promises returned by the browser. For scenarios involving elements that load assets (f.e. `<img>`), browsers fire a `load` event when the element has loaded an asset or assets.
2. Then the user can start to interact with a camera object but not in a standard way (JS only, no compat with tools).

Instead, with a DOM-centric approach, users would be able to simply map their desired state to the DOM (whether attribtues on `<model>` or a new element like `<camera>`). The user does not need to know when the underlying internal camera is ready for rendering, just like the user does not need to know when a `<select>` is ready for rendering: they simply map their state to the DOM, and they expect the display to update. `load` events can be emitted to tell users when assets are ready. `Promise` is not necessary.

The current `getCamera()` essentially creates a new and alternative object model that requires new tooling for all frameworks, yet does not provide a clear benefit that would make this tradeoff worth it.

We can also get creative here, while not veering away from the really great web developer paradigms we already have:

In a further future (and based on my experience making Lume, and knowledge of [A-Frame](https://aframe.io/), [Trois](https://github.com/troisjs/trois) for Vue, [Threlte](https://threlte.xyz/) for Svelte, [react-three-fiber](https://docs.pmnd.rs/react-three-fiber) for React, and other similar libraries for various frameworks), `<camera>` and `<model>` can be decoupled from each other, used in a 3D `<scene>`, implicitly disabling the `<model>`'s camera and instead rendering relative to a user-defined `<scene>` with custom graphics interspersed with any number of `<model>` elements:

```js
<scene fog="linear" fog-color="pink" ...>
  <camera type="perspective" position="..."></camera>
  <model rotation="..." ...></model>
  <model scale="..." ...></model>
  <box ...></box>
  <sphere ...></sphere>
</scene>
```

Such a scene as above is fully compatible with web tools that have evolved over so many years of thought.

Changing the <model> API would be easy, but breaking apps would be extremely difficult. This is why we need to imagine, right now: how do we ensure that the future brings us elements that immediately fit right in with the tools we've worked hard to create?

I would love to help ideate this future. I've started to ideate over at [https://lume.io](https://t.co/FFxzcJue0o), HTML elements for 3D graphics using Three.js underneath.

Aside: the `slot=""` attribute in the first example would explain (per the 
[@w3c](https://twitter.com/w3c)
 community's Extensible Web Manifesto's mantra to "explain existing features", [https://extensiblewebmanifesto.org](https://t.co/66gHgf5s4o)) how the camera works: it could be distributed into an implicit 3D <scene> that the <model> creates when not already inside of an ancestor <scene>. When already in an ancestor <scene>, the model simply provides the 3D asset for rendering within that ancestor scene.

I would love to continue this conversation to ensure that we can reate a set of 3D elements that meet the needs of the vast majority of web developers in the best way possible; in a way that is highly compatible with the existing ecosystem.

[](https://t.co/uVVTsx16Z3)

**Comment by marcoscaceres** [comment]
It's definitely worth exploring the design of the API (e.g., #12 already raised the issue around the promises and #41 and a ton of other issues). 

I think the problem here is that this is conflating what the format provides and what model provides. The `<model>` element, as it stands, it just a simple wrapper for some 3D format (USDZ or gLTF). 

In designing an API, what we need to figure out is exactly what overlap there is between the two formats in accessing the common internals (e.g., they may both have an initial camera position defined in the format itself). We can't really do much more than that, I think... but that's still allows for a lot of stuff, as model-viewer has shown.  

I don't think we want to end up with  `<box ...></box> <sphere ...></sphere>` etc. We'd basically be recreating VRML. 

I guess what would be nice with existing things on the Web is if there was some way to simply light a scene using environmental lighting information and present it in a stereoscopic view (like what VisionOS does), but take away the ability for the page to capture the canvas (for privacy, as the UA would be reflecting the user's environment). 

I don't know if that's possible/feasible, but it seems that's what you'd want here: to continue to support all the legacy content, but just have it be rendered/lit correctly in the new context.

Al the same time, it's nice to just put a `<model>` on the page, just like one does with an `<img>` tag and have it do simple things... for anything more fancy, there is WebGPU, WebGL, WebXR, Canvas etc.

**Comment by zachernuk** [comment]
I totally agree about establishing some manner of composing and manipulating model contents - I'm just conscious that the more that goes into an initial specification, the harder it becomes to agree upon it and the more implementation work must be done before it's complete. To that end, in the interests of finding the _absolute_ MVP, I'm suggesting that we defer any scene inspection until a post-V1. @trusktr Would you be okay to close this?

**Comment by trusktr** [comment]
> I don't think we want to end up with `<box ...></box> <sphere ...></sphere>` etc. We'd basically be recreating VRML.

For the use case of `<model>`, perhaps not. For the use case od 3D in the web, totally.

So, I think it would be ideal if we keep this in mind, even if it is not part of the `<model>` element. The reason is because, then some features can be standardized later. For example, suppose that this is the model MVP:

```html
<model ...>
  <camera type="perspective" position="10 20 30"></camera>
</model>
```

Later on, when browsers have had time to work on their 3d rendering and ironed out all the issues, new ways of making scenes *outside* of the `<model>` element could be introduced, and a feature like `<camera>` could be easily re-used:

```html
<gpuscene>
  <box position="1 2 3"></box>

  <!-- the same camera element people already know how to use: -->
  <camera type="perspective" position="10 20 30"></camera>
</gpuscene>
```

Later on, we could even decide what it means to put a `<model>` inside a 3D scene! For example:

```html
<gpuscene>
  <box position="1 2 3"></box>

  <!-- same model element people already know!: -->
  <model ...></model>

  <!-- the same camera element people already know how to use: -->
  <camera type="perspective" position="10 20 30"></camera>
</gpuscene>
```

Keeping this in mind will help us get to that eventual future which some of us do want.

I know I really do, and I know many people can benefit from it. That's why I'm working on answering the question "what would built-in 3D HTML elements look like if future browsers had them?" over in Lume:

https://lume.io

It would be amazing if Lume (or A-Frame, X3D, etc) all became obsolete projects because browsers one day gave us 3D elements.

**Comment by trusktr** [comment]
I added Copilot summary to the top of the OP, which I'll keep updated over time.

**Comment by zachernuk** [comment]
>  even if it is not part of the <model> element.

I think that's the core question for what is and isn't part of `<model>` - we all agree that there should be a way to create a spatial web, with composable elements, potentially positioned by constraints etc. I also think that it'll be possible to _explore_ aspects of that spatial web with `<model>` - but 1. doesn't need to be Model's primary function and, by consequence 2. should place API obligations on Model to do more than it's supposed to.

This is not solely up to me, but the way I've been thinking of it as representing a _single file of 3D content at a time_, portrayed and manipulated monolithically. In 2024 I revised the control mechanism to remove the idea of `Camera`, with a preference to depict the transform on the Model entity as (very tentatively) the `entityTransform`, [described in the explainer](https://github.com/immersive-web/model-element/blob/main/explainer.md#visual-presentation-control).   

Because Model is intended to act as the simplest closure over 3D content - similar to `<img>` in 2D - it's not expected to provide analogs to more expressive contexts like `<svg>` or `<canvas>`, in either its `CanvasRenderingContext2D` flavor or the mode modern `WebGLRenderingContext` and WebGPU flavors. Those remain highly desirable collections of capability to pursue - I just don't think they have to come through (or start before) statically-provided, monolithic 3D asset content.
